include/ensmallen_bits/lookahead/lookahead_impl.hpp

```diff
@@ -0,0 +1,191 @@
+/**
+ * @file lookahead_impl.hpp
+ * @author Marcus Edel
+ *
+ * Implementation of Lookahead Optimizer: k steps forward, 1 step back.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_LOOKAHEAD_LOOKAHEAD_IMPL_HPP
+#define ENSMALLEN_LOOKAHEAD_LOOKAHEAD_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "lookahead.hpp"
+
+#include <ensmallen_bits/function.hpp>
+
+namespace ens {
+
+template<typename BaseOptimizerType, typename DecayPolicyType>
+inline Lookahead<BaseOptimizerType, DecayPolicyType>::Lookahead(
+    const BaseOptimizerType& baseOptimizer,
+    const double stepSize,
+    const size_t k,
+    const size_t maxIterations,
+    const double tolerance,
+    const DecayPolicyType& decayPolicy,
+    const bool exactObjective) :
+    baseOptimizer(baseOptimizer),
+    stepSize(stepSize),
+    k(k),
+    maxIterations(maxIterations),
+    tolerance(tolerance),
+    decayPolicy(decayPolicy),
+    exactObjective(exactObjective),
+    isInitialized(false)
+{ /* Nothing to do. */ }
+
+//! Optimize the function (minimize).
+template<typename BaseOptimizerType, typename DecayPolicyType>
+template<typename DecomposableFunctionType,
+         typename MatType,
+         typename GradType,
+         typename... CallbackTypes>
+typename std::enable_if<IsArmaType<GradType>::value,
+typename MatType::elem_type>::type
+Lookahead<BaseOptimizerType, DecayPolicyType>::Optimize(
+    DecomposableFunctionType& function,
+    MatType& iterateIn,
+    CallbackTypes&&... callbacks)
+{
+  // Convenience typedefs.
+  typedef typename MatType::elem_type ElemType;
+  typedef typename MatTypeTraits<MatType>::BaseMatType BaseMatType;
+  typedef typename MatTypeTraits<GradType>::BaseMatType BaseGradType;
+
+  typedef Function<DecomposableFunctionType, BaseMatType, BaseGradType>
+      FullFunctionType;
+  FullFunctionType& f(static_cast<FullFunctionType&>(function));
+
+  // The decay policy internally use a templated class so that
+  // we can know MatType and GradType only when Optimize() is called.
+  typedef typename DecayPolicyType::template Policy<BaseMatType, BaseGradType>
+      InstDecayPolicyType;
+
+  // Make sure we have all the methods that we need.
+  traits::CheckDecomposableFunctionTypeAPI<FullFunctionType, BaseMatType,
+      BaseGradType>();
+  RequireFloatingPointType<BaseMatType>();
+  RequireFloatingPointType<BaseGradType>();
+  RequireSameInternalTypes<BaseMatType, BaseGradType>();
+
+  BaseMatType& iterate = (BaseMatType&) iterateIn;
+
+  // Check if the optimizer implements HasMaxIterations() and override the
+  // parameter with k.
+  if (callbacks::traits::HasMaxIterationsSignature<BaseOptimizerType>::value)
```

Comment:
These traits generally don't have to do with the callbacks themselves; if you wanted, maybe we should move them out of the `callbacks` namespace and put them into `function/traits.hpp` or something?  I am not picky either way, and no need to do that in this PR, just wanted to see what you think of the idea.

Reply:
Agreed, not sure why I put them into the callbacks traits.

Reply:
Yeah, it's not a huge deal either way---if it's more convenient I think it's fine to open an issue or do it in a future PR.